### ISSUE TITLE
Expression engine has_role() function can't process implicit roles

### ISSUE THREAD

**Issue by lashae** [symfony/symfony#16096]
I have the following access control rules:

```yaml
    access_control:
        - { path: ^/login$, role: IS_AUTHENTICATED_ANONYMOUSLY }
        - { path: ^/user/register, role: IS_AUTHENTICATED_ANONYMOUSLY }
        - { path: ^/user/resetting, role: IS_AUTHENTICATED_ANONYMOUSLY }
        - { path: ^/admin, allow_if: "has_role('ROLE_ADMIN') and has_role('IS_AUTHENTICATED_FULLY')" }
        - { path: ^/, role: IS_AUTHENTICATED_REMEMBERED }
```

4th access control rule `has_role('ROLE_ADMIN') and has_role('IS_AUTHENTICATED_FULLY')` doesn't work as expected.

If I replace `has_role('IS_AUTHENTICATED_FULLY')` with `is_fully_authenticated()` and revise access control rules as follows:

```yaml
    access_control:
        - { path: ^/login$, role: IS_AUTHENTICATED_ANONYMOUSLY }
        - { path: ^/user/register, role: IS_AUTHENTICATED_ANONYMOUSLY }
        - { path: ^/user/resetting, role: IS_AUTHENTICATED_ANONYMOUSLY }
        - { path: ^/admin, allow_if: "has_role('ROLE_ADMIN') and is_fully_authenticated()" }
        - { path: ^/, role: IS_AUTHENTICATED_REMEMBERED }
```

It works as expected. Further debugging yield me that the implicit `IS_AUTHENTICATED_FULLY` role cannot be processed by expression engine's `has_role()`method properly.

If IS_AUTHENTICATED_REMEMBERED and IS_AUTHENTICATED_FULLY are `role`s (they can be used as a `role` setting in access control rules like in the 1st, 2nd, 3rd and 5th rules above), we should be able to use them in `has_role()` function as well otherwise it is so much confusing.

Please refer to my SO post for additional details: http://stackoverflow.com/questions/32906151/symfony2-access-control-with-multiple-role-requirement/32906940

**Comment by stof** [comment]
``IS_AUTHENTICATED_REMEMBERED`` is not a role.

There is 2 different concepts in the Security system:

- permission attributes (what you ask voters to vote on)
- roles, which are strings starting with ``ROLE_``, hold by the user and handled by the RoleVoter when using them as permission attributes.

The actual issue is that ``role`` in the access_control section is not asking for a role, but for a permission attribute.

**Comment by lashae** [comment]
I agree with you @stof. However, if `IS_AUTHENTICATED_REMEMBERED` and `IS_AUTHENTICATED_FULLY` are not roles, then access_control 's should not be configured with the `role` keyword.

At this point, I understand that asking for a change on this issue may cause a serious BC break, however this inconsistency confuses anyone who tries to read/write security configurations literally.

One other issue is that, these two permission attributes are mentioned as roles several times throughout the Symfony documentation.

Just an example:
[Forcing the User to Re-Authenticate before Accessing certain Resources](http://symfony.com/doc/current/cookbook/security/remember_me.html#forcing-the-user-to-re-authenticate-before-accessing-certain-resources)
> If you have the IS_AUTHENTICATED_REMEMBERED role, then you also have the IS_AUTHENTICATED_ANONYMOUSLY role. If you have the IS_AUTHENTICATED_FULLY role, then you also have the other two roles. In other words, these roles represent three levels of increasing "strength" of authentication.

IMHO, this is an issue to be resolved, that way or another.

**Comment by iltar** [comment]
It's quite simple, `has_role()` only works if it's an actual role assigned, which usually means it's prefixed with `ROLE_`. `IS_AUTHENTICATED_*` implies via the is, that it's checking a state of something. I don't see how this is a problem, other than the example containing a wrong check:

```
@Security("has_role('IS_AUTHENTICATED_FULLY')")
```

As far as I know, this should be:
 
```
@Security("is_granted('IS_AUTHENTICATED_FULLY')")
```

If this is indeed incorrect, you could create an issue or pull request on symfony/symfony-docs

**Comment by lashae** [comment]
Iltar, do you understand my points in terms of consistency and confusing developers?

I understand that "there is a way of doing something.", this is reasonable. You and I and a number of people reading the docs and this issue can learn and use SF as it is designed. But fixing  some natural defects will enable users to use SF w/o referring to the documentation.

Literally, the method name is `has_role(...)` and in access control rules you can add such a rule:

    - { path: ^/, role: IS_AUTHENTICATED_REMEMBERED }

So, if this `IS_AUTHENTICATED_REMEMBERED` is not a role, than why do we use it next to `role` keyword. If it is "something like a role", then why one cannot use it with `has_role(...)`?

To summarize, if you call a car and a jet as a "plane". Then, you should be ready to answer questions like "why this car cannot fly?"

**Comment by iltar** [comment]
I think the proper name for role in there would actually be attribute, the whole "roles" part is confusing within the security.

**Comment by lashae** [comment]
I agree with you. However, that correction is a huge BC issue. I think adding support for permission attributes in `has_role(...)` is the only way to fix this confusion so far.

**Comment by iltar** [comment]
@lashae I think you need to do something in here for that, but I disagree with that change as it's nothing more than patching up the symptoms: https://github.com/symfony/symfony/blob/2aeaa22e3171b5e1f10f2cc2a7ff7f6608b65f9d/src/Symfony/Component/Security/Core/Authorization/ExpressionLanguageProvider.php#L51

**Comment by lashae** [comment]
Patching the framework is the easiest part of the resolution. The most important thing is that ***what does the community think about this issue?*** For example, what do you think? Are you convinced that _this is an issue_ or would you vote to just ignore it?

**Comment by iltar** [comment]
I can't remember being confused about it, but that doesn't mean it's not confusing.

**Comment by javiereguiluz** [comment]
I'm closing this because at the end is a question about using the Security component.

@lashae I understand that this may be confusing for some people at first, but we haven't received lots of reports about this in the past.
